packages/network-controller/src/NetworkController.ts

```diff
@@ -896,6 +1156,246 @@ export class NetworkController extends BaseControllerV2<
       };
     });
   }
-}
 
-export default NetworkController;
+  /**
+   * Before accessing or switching the network, the registry of network clients
+   * needs to be populated. Otherwise, `#applyNetworkSelection` and
+   * `getNetworkClients` will throw an error. This method checks to see if the
+   * population step has happened yet, and if not, makes it happen.
+   *
+   * @returns The populated network client registry.
+   */
+  #ensureAutoManagedNetworkClientRegistryPopulated(): AutoManagedNetworkClientRegistry {
+    const autoManagedNetworkClientRegistry =
+      this.#autoManagedNetworkClientRegistry ??
+      this.#createAutoManagedNetworkClientRegistry();
+    this.#autoManagedNetworkClientRegistry = autoManagedNetworkClientRegistry;
+    return autoManagedNetworkClientRegistry;
+  }
+
+  /**
+   * Constructs the registry of network clients based on the set of built-in
+   * networks as well as the custom networks in state.
+   *
+   * @returns The network clients keyed by ID.
+   */
+  #createAutoManagedNetworkClientRegistry(): AutoManagedNetworkClientRegistry {
+    return [
+      ...this.#buildIdentifiedInfuraNetworkClientConfigurations(),
+      ...this.#buildIdentifiedCustomNetworkClientConfigurations(),
+      this.#buildIdentifiedNetworkClientConfigurationFromProviderConfig(),
+    ].reduce(
+      (
+        obj,
+        [networkClientType, networkClientId, networkClientConfiguration],
+      ) => {
+        const autoManagedNetworkClient = createAutoManagedNetworkClient(
+          networkClientConfiguration,
+        );
+        if (networkClientId in obj[networkClientType]) {
+          return obj;
+        }
+        return {
+          ...obj,
+          [networkClientType]: {
+            ...obj[networkClientType],
+            [networkClientId]: autoManagedNetworkClient,
+          },
+        };
+      },
+      {
+        [NetworkClientType.Infura]: {},
+        [NetworkClientType.Custom]: {},
+      } as AutoManagedNetworkClientRegistry,
+    );
```

Comment:
Nit: Better to cast at the end if we're going to use a cast, so that the type during the `reduce` function isn't inaccurate

```suggestion
      },
    ) as AutoManagedNetworkClientRegistry;
```

Avoiding a cast entirely and getting TypeScript to validate that this type for us would require a very different solution. Probably easier to have a separate `reduce` function for Infura and Custom to accomplish that.

Reply:
Sure. Fixed in [6977c44](https://github.com/MetaMask/core/pull/1439/commits/6977c44277fa8e914c860b46751a8678277c85a9).